src/pw/fpga/opencl_utils.c

```diff
@@ -0,0 +1,370 @@
+      return pid;
+    }
+  }
+  free(pids);
+  return NULL;
+}
+
+/******************************************************************************
+ * \brief   returns the list of all devices for the specfic platform
+ * \param   platform id to search for devices 
+ * \param   specific type of device to search for
+ * \param   total number of devices found for the given platform
+ * \retval  array of device ids
+ *****************************************************************************/
+cl_device_id* getTestDevices(cl_platform_id pid, cl_device_type device_type, cl_uint *num_devices) {
+  cl_int status;
+
+  // Query for number of devices
+  status = clGetDeviceIDs(pid, device_type, 0, NULL, num_devices);
+  if(status != CL_SUCCESS){
+    printf("Query for number of devices failed\n");
+    exit(0);
+  }
+
+  //  Based on the number of devices get their device ids
+  cl_device_id *dev_ids = (cl_device_id*) malloc(sizeof(cl_device_id) * (*num_devices));
+  status = clGetDeviceIDs(pid, device_type, *num_devices, dev_ids, NULL);
+  if(status != CL_SUCCESS){
+    printf("Query for device ids failed\n");
+    free(dev_ids);
+    exit(0);
+  }
+  return dev_ids;
+}
+
+/******************************************************************************
+ * \brief   returns the list of all devices for the specfic platform
+ * \param   context created using device
+ * \param   array of devices
+ * \param   number of devices found
+ * \param   size of FFT3d
+ * \retval  created program
+ *****************************************************************************/
+cl_program getProgramWithBinary(cl_context context, const cl_device_id *devices, unsigned num_device, int N[3], char *data_path){
+  char bin_path[500];
+  char *binary;
+  char *binaries[num_device];
+
+  size_t bin_size;
+  cl_int bin_status, status;
+
+  strcpy(bin_path, data_path);
+
+  // Get the full path of the binary
+  if( !getBinaryPath(bin_path, N) ){
+    printf("No paths to the binary found\n");
+    return NULL;
+  }
+
+  // Load binary to character array
+  binary = loadBinary(bin_path, &bin_size);
+  if(binary == NULL){
+    printf("Could not load binary\n");
+    free(binary);
+    return NULL;
+  }
+
+  binaries[0] = binary;
+
+  // Create the program.
+  cl_program program = clCreateProgramWithBinary(context, 1, devices, &bin_size, (const unsigned char **) binaries, &bin_status, &status);
+  if (status != CL_SUCCESS){
+    printf("Query to create program with binary failed\n");
+    free(binary);
+    return NULL;
+  }
+  free(binary);
+  return program;
+}
+
+static bool fileTestExists(char* filename){
+  if( access( filename, R_OK ) != -1 ) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+static bool getBinaryPath(char *path, int N[3]){
+
+#ifdef __PW_FPGA_SP
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_sp/";
+#else
+  char *subpath = "/../fpgabitstream/fft3d/synthesis_dp/";
+#endif
+  
+  switch(N[0]){
+    case 16 : 
+      strcat(path, subpath);
```

Comment:
possible memory corruption here if `path` is not large enough, same for all other uses of `strcat`